src/Microsoft.AspNetCore.Blazor.Razor.Extensions/GenericComponentPass.cs

```diff
@@ -0,0 +1,169 @@
+                    // what to do, whereas the other errors might be confusing.
+                    node.Diagnostics.Add(BlazorDiagnosticFactory.Create_GenericComponentMissingTypeArgument(node.Source, node, missing));
+                }
+
+                var rewriter = new TypeNameRewriter(bindings);
+
+                // Rewrite the component type name
+                node.TypeName = RewriteTypeName(rewriter, node.TypeName);
+
+                foreach (var attribute in node.Attributes)
+                {
+                    if (attribute.BoundAttribute?.IsGenericTypedProperty() ?? false && attribute.TypeName != null)
+                    {
+                        // If we know the type name, then replace any generic type parameter inside it with
+                        // the known types.
+                        attribute.TypeName = RewriteTypeName(rewriter, attribute.TypeName);
+                    }
+                }
+
+                foreach (var childContent in node.ChildContents)
+                {
+                    if (childContent.BoundAttribute?.IsGenericTypedProperty() ?? false && childContent.TypeName != null)
+                    {
+                        // If we know the type name, then replace any generic type parameter inside it with
+                        // the known types.
+                        childContent.TypeName = RewriteTypeName(rewriter, childContent.TypeName);
+                    }
+                }
+            }
+
+            private string RewriteTypeName(TypeNameRewriter rewriter, string typeName)
+            {
+                var parsed = SyntaxFactory.ParseTypeName(typeName);
+                var rewritten = (TypeSyntax)rewriter.Visit(parsed);
+                return rewritten.ToFullString();
+            }
+
+            private string GetContent(ComponentTypeArgumentExtensionNode node)
+            {
+                return string.Join(string.Empty, node.FindDescendantNodes<IntermediateToken>().Where(t => t.IsCSharp).Select(t => t.Content));
+            }
+        }
+
+        private class TypeNameRewriter : CSharpSyntaxRewriter
+        {
+            private readonly Dictionary<string, Binding> _bindings;
+
+            public TypeNameRewriter(Dictionary<string, Binding> bindings)
+            {
+                _bindings = bindings;
+            }
+
+            public override SyntaxNode Visit(SyntaxNode node)
+            {
+                // We can handle a single IdentifierNameSyntax at the top level (like 'TItem)
+                // OR a GenericNameSyntax recursively (like `List<T>`)
+                if (node is IdentifierNameSyntax identifier)
+                {
+                    if (_bindings.TryGetValue(identifier.Identifier.Text, out var binding))
+                    {
+                        // If we don't have a valid replacement, use object. This will make the code at least reasonable
+                        // compared to leaving the type parameter in place.
```

Comment:
I think that's a good plan. It means in a lot of cases there *won't* be any other errors, and people will just not get any intellisense prompts on their `context` variable (besides `ToString()`, `GetHashCode()`, etc.), which is ideal.

Technically there's the case that the component declares `where T:struct`, triggering a different error, but that would be very rare and I don't think worth digging further into.

Reply:
yeah, I was optimizing here to try and make the error that we create be the only one the user sees. You'd also have issues if the generic type was intended to be a delegate type, since lambdas aren't implicitly assignable to object, and MethodGroup to delegate conversion isn't valid in that context either.